Add rel="noopener" to output of wp_list_bookmarks() when target is set.

costdev opened the pull request:
To help aid in hardening WordPress and the web as a whole, it became a best practice to add "noopener" to the `rel` attribute for links that have a target.

This updates `wp_list_bookmarks()` so that it now adds "noopener" to the `rel` attribute when the link's target is set.

Trac ticket: https://core.trac.wordpress.org/ticket/53839

Commit:
Add patch and unit tests.

hellofromtonya commented on src/wp-includes/bookmark-template.php:
```diff
@@ -102,13 +102,27 @@ function _walk_bookmarks( $bookmarks, $args = '' ) {
+
 		$target = $bookmark->link_target;
 		if ( '' !== $target ) {
+			if ( ! empty( $rel ) ) {
```
```suggestion
			if ( is_string( $rel ) && '' !== $rel ) {
```

hellofromtonya committed:
Guard PHP native `explode()` to ensure input is a string.

hellofromtonya commented on src/wp-includes/bookmark-template.php:
```diff
@@ -102,13 +102,27 @@ function _walk_bookmarks( $bookmarks, $args = '' ) {
+			if ( is_string( $rel ) && '' !== $rel ) {
+				$rels   = explode( ' ', $rel );
+
+				if ( false === in_array( 'noopener', $rels, true ) ) {
+					$rels[] = 'noopener';
+				}
+				$rel    = implode( ' ', $rels );
```
Micro-optimization: Dop the `implode()` only when adding `noopener` to the list of rels.
```suggestion
				if ( in_array( 'noopener', $rels, true ) ) {
					$rels[] = 'noopener';
					$rel     = implode( ' ', $rels );
				}
```

hellofromtonya committed:
Micro-optimization: do `implode()` if `'noopener'` added.

hellofromtonya commented on src/wp-includes/bookmark-template.php:
```diff
@@ -102,13 +102,27 @@ function _walk_bookmarks( $bookmarks, $args = '' ) {
 		$target = $bookmark->link_target;
 		if ( '' !== $target ) {
+			if ( is_string( $rel ) && '' !== $rel ) {
+				$rels   = explode( ' ', $rel );
+
+				if ( in_array( 'noopener', $rels, true ) ) {
+					$rels[] = 'noopener';
+					$rel     = implode( ' ', $rels );
+				}
```
Replace with less expensive processing:

```suggestion
				if ( ! str_contains( $rels, 'noopener' ) ) {
					$rel = trim( $rel ) . ' noopener';
				}
```

A polyfill was added today to Core for using `str_contains()` (i.e. a new PHP 8.0 function). This function eliminates parsing into an array, searching the array, and then converting the array back into a string. Less memory and processing.

hellofromtonya committed:
Use `str_contains()` to search for `noopener`.

A polyfill was added today to Core for using str_contains() (i.e. a new PHP 8.0 function). This function eliminates parsing into an array, searching the array, and then converting the array back into a string. Less memory and processing.

hellofromtonya reviewed (approved):
Ready for commit 👍

hellofromtonya commented on src/wp-includes/bookmark-template.php:
```diff
@@ -102,13 +102,24 @@ function _walk_bookmarks( $bookmarks, $args = '' ) {
 		$target = $bookmark->link_target;
 		if ( '' !== $target ) {
+			if ( is_string( $rel ) && '' !== $rel ) {
+				if ( ! str_contains( $rels, 'noopener' ) ) {
```
```suggestion
				if ( ! str_contains( $rel, 'noopener' ) ) {
```
🤦‍♀️ doh on my part.

hellofromtonya committed:
Fix (doh) typo.

hellofromtonya commented:
Committed via changeset https://core.trac.wordpress.org/changeset/52061.
